### PR TITLE
Aave v3 updates

### PR DESCRIPTION
**Pull request by shriyatyagii**
New functions added on Aave V3 - 

1. `depositWithoutCollateral`
2. `borrowOnBehalfOf`
3. `approveDelegation`
4. `paybackOnBehalfOf`

Deployed Implementations - 
Arbitrum - [0xbA5B64a60e6d4E983c1A2B34913d51c7F74E2017](https://arbiscan.io/address/0xbA5B64a60e6d4E983c1A2B34913d51c7F74E2017#code)
Avalanche - [0x55D11D1076abFAa85499744633cd609023Fe47ef](https://snowtrace.io/address/0x55D11D1076abFAa85499744633cd609023Fe47ef#code)
Fantom - [0xB99B9d50366c2a00c61415948953A1429bBD4450](https://ftmscan.com/address/0xB99B9d50366c2a00c61415948953A1429bBD4450#code)
Optimism - [0x22ed23Cc6EFf065AfDb7D5fF0CBf6886fd19aee1](https://optimistic.etherscan.io/address/0x22ed23Cc6EFf065AfDb7D5fF0CBf6886fd19aee1#code)
Polygon - [0xEe97AE0d5943ff419254191E2361A81cDf7F50E1](https://polygonscan.com/address/0xEe97AE0d5943ff419254191E2361A81cDf7F50E1#code)

- [x] Review and approve @thrilok209 @pradyuman-verma 
- [x] deploy and enable
- [x] update SDK https://github.com/Instadapp/dsa-connect/pull/161
- [x] update docs

### CODE REVIEW

**Review by pradyuman-verma — DISMISSED** [review]

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 1]
```diff
@@ -316,6 +404,32 @@ abstract contract AaveResolver is Events, Helpers {
...
+
+	/**
+	 * @dev Approve Delegation
+	 * @notice Gives approval to delegate debt tokens
+	 * @param token The address of token
+	 * @param rateMode The type of borrow debt
+	 * @param delegateTo The address to whom the user is delegating
+	 * @param amount The amount 
+	 */
+	function approveDelegation(address token, uint16 rateMode, address delegateTo, uint256 amount)
```
Comment (thrilok209) [inline contracts/arbitrum/connectors/aave/v3/main.sol]: @shriyatyagii I think we can add getId and setId too? Just incase

Comment (thrilok209) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
	function approveDelegation(address token, , uint256 amount, uint256 rateMode, address delegateTo)
```

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 2]
```diff
@@ -139,6 +189,44 @@ abstract contract AaveResolver is Events, Helpers {
...
+
+		bool isEth = token == ethAddr;
+		address _token = isEth ? wethAddr : token;
+
+		aave.borrow(_token, _amt, rateMode, referralCode, onBehalfOf);
+		convertEthToWeth(isEth, TokenInterface(_token), _amt);
+
+		setUint(setId, _amt);
+
+		_eventName = "LogBorrow(address,uint256,uint256,uint256,uint256)";
```
Comment (thrilok209) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
		_eventName = "LogBorrowOnBehalfOf(address,uint256,uint256,address,uint256,uint256)";
```

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 3]
```diff
@@ -63,6 +63,56 @@ abstract contract AaveResolver is Events, Helpers {
...
+
+		aave.supply(_token, _amt, address(this), referralCode);
+
+		if (getCollateralBalance(_token) > 0 && getIsColl(token)) {
+			aave.setUserUseReserveAsCollateral(_token, false);
+		}
+
+		setUint(setId, _amt);
+
+		_eventName = "LogDeposit(address,uint256,uint256,uint256)";
```
Comment (thrilok209) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
		_eventName = "LogDepositWithoutCollateral(address,uint256,uint256,uint256)";
```

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 4]
```diff
@@ -316,6 +404,32 @@ abstract contract AaveResolver is Events, Helpers {
...
+		address _dToken = getDTokenAddr(_token, rateMode);
+		DTokenInterface(_dToken).approveDelegation(delegateTo, amount);
+
+		_eventName = "LogApproveDelegation(address,uint16,address,uint256)";
+		_eventParam = abi.encode(token, rateMode, delegateTo, amount);
+
+	}
 }
 
 contract ConnectV2AaveV3Arbitrum is AaveResolver {
```
Comment (thrilok209) [inline contracts/arbitrum/connectors/aave/v3/main.sol:491]: image
Need to change to `AaveV3-v1.2`

Comment (shriyatyagii) [reply contracts/arbitrum/connectors/aave/v3/main.sol:491]: Got it

File: contracts/arbitrum/connectors/aave/v3/main.sol (hunk 1 above)
Comment (shriyatyagii) [reply contracts/arbitrum/connectors/aave/v3/main.sol]: Added

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 5]
```diff
@@ -63,6 +63,56 @@ abstract contract AaveResolver is Events, Helpers {
 		_eventParam = abi.encode(token, _amt, getId, setId);
 	}
 
+	/**
+	 * @dev Deposit ETH/ERC20_Token.
```
Comment (thrilok209) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
	 * @dev Deposit ETH/ERC20_Token without collateral
```

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 6]
```diff
@@ -63,6 +63,56 @@ abstract contract AaveResolver is Events, Helpers {
...
+	 * @notice Deposit a token to Aave v3 for lending / collaterization.
```
Comment (thrilok209) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
	 * @notice Deposit a token to Aave v3 without enabling it as collateral.
```

File: contracts/polygon/connectors/aave/v3/main.sol [hunk 7]
```diff
@@ -313,6 +313,51 @@ abstract contract AaveResolver is Events, Helpers {
 		_eventParam = abi.encode(token, _amt, rateMode, getId, setId);
 	}
 
+	/**
+	 * @dev Payback borrowed matic/ERC20_Token on behalf os a user.
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3/main.sol]: on behalf of a user

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 8]
```diff
@@ -139,6 +189,44 @@ abstract contract AaveResolver is Events, Helpers {
 		_eventParam = abi.encode(token, _amt, rateMode, getId, setId);
 	}
 
+	/**
+	 * @dev Borrow ETH/ERC20_Token on behalf of a user.
+	 * @notice Borrow a token using Aave v3 on behalf of a user
+	 * @param token The address of the token to borrow.(For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param amt The amount of the token to borrow.
+	 * @param rateMode The type of borrow debt. (For Stable: 1, Variable: 2)
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
	 * @param rateMode The type of debt. (For Stable: 1, Variable: 2)
```

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 9]
```diff
@@ -316,8 +404,45 @@ abstract contract AaveResolver is Events, Helpers {
 		_eventName = "LogSetUserEMode(uint8)";
 		_eventParam = abi.encode(categoryId);
 	}
+
+	/**
+	 * @dev Approve Delegation
+	 * @notice Gives approval to delegate debt tokens
+	 * @param token The address of token
+	 * @param amount The amount
+	 * @param rateMode The type of borrow debt
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
	 * @param rateMode The type of debt
```

File: contracts/arbitrum/connectors/aave/v3/main.sol [hunk 10]
```diff
@@ -316,8 +404,45 @@ abstract contract AaveResolver is Events, Helpers {
...
+	 * @param delegateTo The address to whom the user is delegating
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of tokens deposited.
```
Comment (pradyuman-verma) [inline contracts/arbitrum/connectors/aave/v3/main.sol]:
```suggestion
	 * @param setId ID stores the amount of tokens delegated.
```

**Comment by shriyatyagii** [comment]
**Deployed on Arbitrum**
Name - `AAVE-V3-A`
Arbitrum - [0xbA5B64a60e6d4E983c1A2B34913d51c7F74E2017](https://arbiscan.io/address/0xbA5B64a60e6d4E983c1A2B34913d51c7F74E2017#code)
Tx - [Arbiscan](https://arbiscan.io/tx/0x6a8621dfef2f477eac218c7436e3ffdfa17b57ed983dbbd42fdb992c3bd9f537#eventlog)

**Deployed on Avalanche**
Name - `AAVE-V3-A`
Avalanche - [0x55D11D1076abFAa85499744633cd609023Fe47ef](https://snowtrace.io/address/0x55D11D1076abFAa85499744633cd609023Fe47ef#code)
Tx - [Snowtrace](https://snowtrace.io/tx/0x960e84f0aca7ecb547a1f3de27538c6c8ac1ffbcdc894108404da7877e2a0eab)

**Deployed on Fantom**
Name - `AAVE-V3-A`
Fantom - [0xB99B9d50366c2a00c61415948953A1429bBD4450](https://ftmscan.com/address/0xB99B9d50366c2a00c61415948953A1429bBD4450#code)
Tx- [FTMScan](https://ftmscan.com/tx/0xfbcba5d8468cf20f5ec3b5ee7bf0dbc771b2426c2d54c304b82348c19c51fd4f)

**Deployed on Optimism**
Name - `AAVE-V3-A`
Optimism - [0x22ed23Cc6EFf065AfDb7D5fF0CBf6886fd19aee1](https://optimistic.etherscan.io/address/0x22ed23Cc6EFf065AfDb7D5fF0CBf6886fd19aee1#code)
Tx - [Optimistic](https://optimistic.etherscan.io/tx/0x43fce94290094519abb16fb4addf6869fc680b1a5c5945ff9496cbf3ced840d4)

**Deployed on Polygon**
Name - `AAVE-V3-A`
Polygon - [0xEe97AE0d5943ff419254191E2361A81cDf7F50E1](https://polygonscan.com/address/0xEe97AE0d5943ff419254191E2361A81cDf7F50E1#code)
Tx - [Polyscan](https://polygonscan.com/tx/0xe7a8509e6e6510093f6bdf5ef6ff3b77b7701b9a61671265b7e6e36e500ea671)

**Review by pradyuman-verma — DISMISSED** [review]

**Review by thrilok209 — APPROVED** [review]

**Pull request merged**